### ISSUE TITLE
How to solve “type is partially unknown” warning?

### ISSUE THREAD

**Issue by Spenhouet** [microsoft/pyright#698]
**Describe the bug**
Pyright is unable to infer generic class type.

**To Reproduce**
Use pytorchs Dataset or DataLoader classes.

**Expected behavior**
Pyright infers the type of the DataLoader from the Dataset parameter as defined by the type stub:

```
class DataLoader(Generic[T_co]):
    dataset: Dataset[T_co]

    @overload
    def __init__(self, dataset: Dataset[T_co], ...
```

Pyright infers the type of the Dataset from the` __getitem__` return type as defined by the type stub:

```
class Dataset(Generic[T_co]):
    def __getitem__(self, index: int) -> T_co: ...
```

**MWE**

```
from torch.utils.data import Dataset, DataLoader


class TestDataset(Dataset):

    def __init__(self):
        pass

    def __len__(self) -> int:
        return 100

    def __getitem__(self, index: int) -> int:
        return index


def get_dataset() -> Dataset:
    dataset = TestDataset()
    return dataset


def get_dataloader() -> DataLoader:
    dataset = get_dataset()
    dataloader = DataLoader(dataset=dataset)
    return dataloader
```

**Pyright warnings**

```
def get_dataset() -> Dataset:
                     ^^^^^^^^
                     Declared return type, "Dataset[Unknown]", is partially unknown - Pyright (reportUnknownVariableType)
```

```
def get_dataloader() -> DataLoader:
                        ^^^^^^^^^^
                        Declared return type, "DataLoader[Unknown]", is partially unknown - Pyright (reportUnknownVariableType)
```

```
    dataset = get_dataset()
    ^^^^^^^
    Type of "dataset" is partially unknown
      Type of "dataset" is "Dataset[Unknown]" - Pyright (reportUnknownVariableType)
```

```
    dataloader = DataLoader(dataset=dataset)
    ^^^^^^^^^
    Type of "dataloader" is partially unknown
      Type of "dataloader" is "DataLoader[Unknown]" - Pyright (reportUnknownVariableType)
```

```
    dataloader = DataLoader(dataset=dataset)
                                    ^^^^^^^
                                    Argument type is partially unknown
                                      Argument corresponds to parameter "dataset" in function "__init__"
                                        Argument type is "Dataset[Unknown]" - Pyright (reportUnknownArgumentType)
```

```
    return dataloader
           ^^^^^^^^^
           Return type, "DataLoader[Unknown]", is partially unknown - Pyright (reportUnknownVariableType)
```

**VS Code extension or command-line**
I'm using the strict setting for pyright:

`"pyright.typeCheckingMode": "strict",`

**Additional context**
I thought this was an error on my part and had opened a stackoverflow question: https://stackoverflow.com/questions/61997937
Since there are no replies, I'm not sure if this is a bug.

**Comment by erictraut** [comment]
Did you generate this type stub, or did you download an existing type stub?

When you use strict settings in Pyright, the "reportUnknownVariableType" diagnostic rule is enabled. It emits an error when a type is fully or partially unknown. In this particular case, the type arguments are not specified in your type stub and are therefore unknown. For example, `Dataset` is a generic class that takes a single type argument (e.g. `DataSet[str]`), but the type stub is specifying `Dataset` without specifying the type argument. If the type argument can truly be any type, then the stub should specify `Dataset[Any]`. If the type is available as a TypeVar, it should be provided (e.g. `Dataset[_T]`).

If you don't want to (or are unable to) fix your type stub, you have a few options:
1. Change the severity level of the "reportUnknownVariableType" diagnostic rule to "warning" or "none" in the python.analysis.diagnosticSeverityOverrides setting to silence it.
2. Silence this rule on a file-by-file basis by adding the following comment at the top of the file:
  # pyright: reportUnknownVariableType=false

**Comment by Spenhouet** [comment]
> Did you generate this type stub, or did you download an existing type stub?

No, these type definitions are supplied by the pytorch package.

dataset.pyi:

```
T_co = TypeVar('T_co', covariant=True)
T = TypeVar('T')
class Dataset(Generic[T_co]):
    def __getitem__(self, index: int) -> T_co: ...
    def __len__(self) -> int: ...
    def __add__(self, other: T_co) -> 'ConcatDataset[T_co]': ...
```

dataloader.pyi:

```
T_co = TypeVar('T_co', covariant=True)
T = TypeVar('T')
_worker_init_fn_t = Callable[[int], None]
_collate_fn_t = Callable[[List[T]], Any]

class DataLoader(Generic[T_co]):
    dataset: Dataset[T_co]
    batch_size: int
    num_workers: int
    pin_memory: bool
    drop_last: bool
    timeout: float

    @overload
    def __init__(self, dataset: Dataset[T_co], batch_size: int=..., shuffle: bool=...,
                 sampler: Optional[Sampler[int]]=..., num_workers: int=..., collate_fn: _collate_fn_t=...,
                 pin_memory: bool=..., drop_last: bool=..., timeout: float=...,
                 worker_init_fn: _worker_init_fn_t=...) -> None: ...
    @overload
    def __init__(self, dataset: Dataset[T_co], batch_sampler: Optional[Sampler[Sequence[int]]]=...,
                 num_workers: int=..., collate_fn: _collate_fn_t=..., pin_memory: bool=..., timeout: float=...,
                 worker_init_fn: _worker_init_fn_t=...) -> None: ...

    def __len__(self) -> int: ...

    def __iter__(self) -> '_BaseDataLoaderIter':...
```

> For example, Dataset is a generic class that takes a single type argument (e.g. DataSet[str]), but the type stub is specifying Dataset without specifying the type argument.

I'm not sure what you are suggesting. As explained in my original post the generic type can be inferred. The type is supplied via the respective parameter and return type as defined by the type signature.

If you suggest that I should manually supply a type to Dataset and DataLoader like this:

```
from torch.utils.data import Dataset, DataLoader


class TestDataset(Dataset):

    def __init__(self):
        pass

    def __len__(self) -> int:
        return 100

    def __getitem__(self, index: int) -> int:
        return index


def get_dataset() -> Dataset[int]:
    dataset = TestDataset()
    return dataset


def get_dataloader() -> DataLoader[int]:
    dataset = get_dataset()
    dataloader = DataLoader(dataset=dataset, batch_size=20)
    return dataloader


def do_something():
    dataloader = get_dataloader()

    for index, elem in enumerate(dataloader):
        print(index, elem)


do_something()
```

This is not valid in python as I also mentioned in the linked stackoverflow issue.

If you execute the above you will get a TypeError at:

```
def get_dataset() -> Dataset[int]:
                     ^^^^^^^^^^
                     Exception has occurred: TypeError
                    'type' object is not subscriptable
                      File "C:\Users\spenh\Downloads\pyright_infer_bug\test.py", line 16, in <module>
                        def get_dataset() -> Dataset[int]:
```

The same will happen for DataLoader.

I don't want to suppress the warnings. I would like to be able to implement the above code correctly with typing. As I see it, this is either possible and then I'm asking how or this is currently not possible due to some bug or missing functionality.

I zipped you the above code with the strict settings:
[pyright_infer_types.zip](https://github.com/microsoft/pyright/files/4700280/pyright_infer_types.zip)
Maybe you could take a look at it.

**Comment by erictraut** [comment]
>This is not valid in python
It is legal Python syntax, but some generic classes unfortunately don't handle subscripting at runtime and generate an exception. That's apparently the case for these pytorch classes. A workaround is to surround the type annotation in quotes, like this:

```
def get_dataset() -> "Dataset[int]":
```

I wish there were a way for type checkers to know whether a class supported subscripting at runtime. Pyright could then detect the problem and emit an error if you didn't use quotes. Unfortunately, this information isn't provided in any form that a type checker has access to.

**Comment by erictraut** [comment]
I should also point out that Pyright will attempt to infer return types if you don't provide a return type annotation for your function or method. But if you provide a return type annotation, it assumes that it should use that type. It does not attempt to use part of the annotated type and fill in missing parts, such as unspecified type arguments.

**Comment by erictraut** [comment]
Another point of potential interest to you. It's unfortunate that the Python interpreter evaluates type annotations. After all, they are not needed or used at runtime. That's why you're seeing the subscript exception. There is a move to fix this unfortunate behavior. It's called "postponed evaluations of type annotations", and it's documented in [PEP 563](https://www.python.org/dev/peps/pep-0563/). If you're using Python 3.7 or newer, you can enable this functionality in your code by adding the following line at the top of your file:

```
from __future__ import annotations
```

Once you add this, you can avoid adding the quotes around your return type annotations, and your code will run without crashing.

**Comment by Spenhouet** [comment]
> It is legal Python syntax

Yes, sorry, meant to say "it does fail" and not that it is generally invalid syntax.

> A workaround is to surround the type annotation in quotes, like this

I did not know I could do that. Yes, I think that workaround is sufficient.

>  It does not attempt to use part of the annotated type and fill in missing parts

Ah, that makes sense.

>  If you're using Python 3.7 or newer, you can enable this functionality

Yes, we are. That is indeed interesting. Thanks for the hint. Adding on to this: I also needed to disable the pylint rule E1136 for pylint to not warn about "unsubscriptable-object".

Thank you for all this information :)
